### PR TITLE
Added Profile Screen

### PR DESCRIPTION
**Pull request by mlchct**
## Summary:
Added Profile Screen

## Changes:
* Changed User model to have description
* Added swiftUI Profile View and ViewModel

## Screenshots

[Uploading Simulator Screenshot - Iphone 13 - 2024-04-25 at 18.04.06.png…]()

### CODE REVIEW

File: Instagram-like-trainee/Coordinators/TabsCoordinators/HomePageCoordinator.swift [hunk 1]
```diff
@@ -33,18 +36,31 @@ final class HomePageCoordinator: HomeCoordinator {
...
+    func didPressProfile(userId: Int) {
+        let profileViewModel = ProfileViewModel(coordinator: self, id: userId, jsonService: jsonService)
+        let view = ProfileView(viewModel: profileViewModel)
+        let hostingController = UIHostingController(rootView: view)
+        navigationController.pushViewController(hostingController, animated: true)
+    }
+
+    func closeProfile() {
+        navigationController.popViewController(animated: true)
+
```
[the remark this answers is not in the document]
Comment (mlchct) [reply Instagram-like-trainee/Coordinators/TabsCoordinators/HomePageCoordinator.swift:61]: Done

File: Instagram-like-trainee/Screens/HomePage/View/CollectionViewCells/PostCell/PostHeaderView.swift [hunk 2]
```diff
@@ -85,4 +91,10 @@ final class PostHeaderView: UIView {
             moreImageView.heightAnchor.constraint(equalToConstant: 24)
         ])
     }
+
+    @objc func openProfile(_ sender: UITapGestureRecognizer) {
+        guard let didPressProfile, let postId else { return}
```
[the remark this answers is not in the document]
Comment (mlchct) [reply Instagram-like-trainee/Screens/HomePage/View/CollectionViewCells/PostCell/PostHeaderView.swift:96]: Done

File: Instagram-like-trainee/Screens/HomePage/View/HomePageViewController.swift [hunk 3]
```diff
@@ -89,9 +89,7 @@ final class HomePageViewController: UIViewController {
     private func createLayout() -> UICollectionViewCompositionalLayout {
         let supplementaryViews = [createHeader()]
 
-        let layout = UICollectionViewCompositionalLayout {
-            sectionIndex,
-            _ -> NSCollectionLayoutSection? in
+        let layout = UICollectionViewCompositionalLayout {sectionIndex, _ -> NSCollectionLayoutSection? in
```
[the remark this answers is not in the document]
Comment (mlchct) [reply Instagram-like-trainee/Screens/HomePage/View/HomePageViewController.swift:92]: Done

File: Instagram-like-trainee/Screens/HomePage/View/HomePageViewController.swift [hunk 4]
```diff
@@ -202,10 +200,14 @@ extension HomePageViewController:
...
                         postHeaderImageURL: viewModel.getUserData(
                             id: post.userId
-                        )?.profileImage ?? URL(string: "")!,
+                        )?.profileImage ?? URL(
+                            string: ""
+                        )!,
                         postUserName: viewModel.getUserData(
                             id: post.userId
-                        )?.name ?? ""
+                        )?.name ?? "",
```
[the remark this answers is not in the document]
Comment (mlchct) [reply Instagram-like-trainee/Screens/HomePage/View/HomePageViewController.swift:208]: Done

File: Instagram-like-trainee/Screens/HomePage/View/HomePageViewController.swift [hunk 5]
```diff
@@ -202,10 +200,14 @@ extension HomePageViewController:
                         postImageURL: post.firstPhotoURL,
                         postHeaderImageURL: viewModel.getUserData(
                             id: post.userId
-                        )?.profileImage ?? URL(string: "")!,
+                        )?.profileImage ?? URL(
+                            string: ""
+                        )!,
```
[the remark this answers is not in the document]
Comment (mlchct) [reply Instagram-like-trainee/Screens/HomePage/View/HomePageViewController.swift:205]: Done

File: Instagram-like-trainee/Screens/ProfilePage/View/ProfileView.swift [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    var body: some View {
+        ScrollView {
+            VStack {
+                NavigationView(profileName: profileData.profileName, closeProfile: closeProfile)
+                HeaderView(profileImage: profileData.profileImage)
+                DescriptionView(profileName: profileData.profileName, description: profileData.description)
+                HStack(spacing: 5) {
+                    createButton(name: "Follow", color: .white)
+                    createButton(name: "Message", color: .black)
+                    createButton(name: "Email", color: .black)
```
[the remark this answers is not in the document]
Comment (mlchct) [reply Instagram-like-trainee/Screens/ProfilePage/View/ProfileView.swift:37]: Done

File: Instagram-like-trainee/Screens/ProfilePage/View/SubViews/HeaderView.swift [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+                        endPoint: .trailing), lineWidth: 3)
+                    .frame(width: 88, height: 88)
+            )
+            Spacer()
+            crateSectionItem(type: .posts(value: 123))
+            Spacer()
+            crateSectionItem(type: .followers(value: 456))
+            Spacer()
+            crateSectionItem(type: .following(value: 798))
+        }.padding()
```
[the remark this answers is not in the document]
Comment (mlchct) [reply Instagram-like-trainee/Screens/ProfilePage/View/SubViews/HeaderView.swift:36]: Done

File: Instagram-like-trainee/Screens/ProfilePage/View/ProfileView.swift [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+            if let profileData = viewModel.data {
+                MainView(profileData: profileData, closeProfile: viewModel.closeProfile)
+            } else {
+                ProgressView("Loading...")
+            }
+        }
+    }
+}
+
+struct MainView: View {
```
[the remark this answers is not in the document]
Comment (mlchct) [reply Instagram-like-trainee/Screens/ProfilePage/View/ProfileView.swift:24]: Done

File: Instagram-like-trainee/Screens/ProfilePage/View/ProfileView.swift [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+                NavigationView(profileName: profileData.profileName, closeProfile: closeProfile)
+                HeaderView(profileImage: profileData.profileImage)
+                DescriptionView(profileName: profileData.profileName, description: profileData.description)
+                HStack(spacing: 5) {
+                    createButton(name: "Follow", color: .white)
+                    createButton(name: "Message", color: .black)
+                    createButton(name: "Email", color: .black)
+                }.padding(3)
+                PostFeedView(posts: profileData.posts)
+
```
[the remark this answers is not in the document]
Comment (mlchct) [reply Instagram-like-trainee/Screens/ProfilePage/View/ProfileView.swift:40]: Done

File: Instagram-like-trainee/Screens/ProfilePage/View/ProfileView.swift [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+            }
+        }
+    }
+
+    func createButton(name: String, color: Color) -> some View {
+        Button {
+
```
[the remark this answers is not in the document]
Comment (mlchct) [reply Instagram-like-trainee/Screens/ProfilePage/View/ProfileView.swift:47]: Done

File: Instagram-like-trainee/Screens/ProfilePage/View/ProfileView.swift [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+        }label: {
```
[the remark this answers is not in the document]
Comment (mlchct) [reply Instagram-like-trainee/Screens/ProfilePage/View/ProfileView.swift:48]: Done

File: Instagram-like-trainee/Screens/ProfilePage/View/SubViews/DescriptionView.swift [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+            VStack(alignment: .leading) {
+                Text(profileName.split(separator: " ").first.map(String.init) ?? "")
+                    .font(.headline)
+                    .multilineTextAlignment(.leading)
+                Text(description)
+                    .font(.subheadline)
+                    .multilineTextAlignment(.leading)
+            }
+            Spacer()
+        }.padding([.leading, .trailing], 10)
```
[the remark this answers is not in the document]
Comment (mlchct) [reply Instagram-like-trainee/Screens/ProfilePage/View/SubViews/DescriptionView.swift:25]: Done

File: Instagram-like-trainee/Screens/ProfilePage/View/SubViews/HeaderView.swift [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+        HStack {
+            AsyncImage(url: profileImage) {image in
+                image
+                    .resizable()
+                    .frame(width: 80, height: 80)
+                    .aspectRatio(contentMode: .fit)
+                    .clipShape(Circle())
+            } placeholder: {
+                ProgressView()
+            }.overlay(
```
[the remark this answers is not in the document]
Comment (mlchct) [reply Instagram-like-trainee/Screens/ProfilePage/View/SubViews/HeaderView.swift:22]: Done

File: Instagram-like-trainee/Screens/ProfilePage/View/SubViews/HeaderView.swift [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+//  Created by Mikhail Kalatsei on 25/04/2024.
+//
+
+import SwiftUI
+
+struct HeaderView: View {
+    var profileImage: URL
+    var body: some View {
+        HStack {
+            AsyncImage(url: profileImage) {image in
```
[the remark this answers is not in the document]
Comment (mlchct) [reply Instagram-like-trainee/Screens/ProfilePage/View/SubViews/HeaderView.swift:14]: Done

File: Instagram-like-trainee/Screens/ProfilePage/View/SubViews/PostFeedView.swift [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+        GridItem(.adaptive(minimum: 120)),
+        GridItem(.adaptive(minimum: 120))
+    ]
+
+    var body: some View {
+        LazyVGrid(columns: adaptiveColumn,
+                  spacing: 3,
+                  content: {
+            ForEach(posts) {post in
+                PostView(name: post.content.first ?? URL(string: "")!)
```
[the remark this answers is not in the document]
Comment (mlchct) [reply Instagram-like-trainee/Screens/ProfilePage/View/SubViews/PostFeedView.swift:24]: Done

File: Instagram-like-trainee/Screens/ProfilePage/View/SubViews/PostFeedView.swift [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+            }
+        }).padding(5)
```
[the remark this answers is not in the document]
Comment (mlchct) [reply Instagram-like-trainee/Screens/ProfilePage/View/SubViews/PostFeedView.swift:26]: Done

File: Instagram-like-trainee/Screens/ProfilePage/View/SubViews/PostFeedView.swift [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+        GridItem(.adaptive(minimum: 120)),
+        GridItem(.adaptive(minimum: 120)),
+        GridItem(.adaptive(minimum: 120))
+    ]
+
+    var body: some View {
+        LazyVGrid(columns: adaptiveColumn,
+                  spacing: 3,
+                  content: {
+            ForEach(posts) {post in
```
[the remark this answers is not in the document]
Comment (mlchct) [reply Instagram-like-trainee/Screens/ProfilePage/View/SubViews/PostFeedView.swift:23]: Done

**Review by e-garbuzov-wise — APPROVED** [review]

File: Instagram-like-trainee/Screens/ProfilePage/View/SubViews/MainView.swift [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+        } label: {
+            Text(name)
+                .font(.subheadline)
+                .bold()
+        }
+        .frame(maxWidth: .infinity, minHeight: 30, maxHeight: 35)
+        .background(color == .white ? .blue : .profileButtonBackground)
+        .cornerRadius(6)
+        .foregroundStyle(color)
+
```
[the remark this answers is not in the document]
Comment (mlchct) [reply Instagram-like-trainee/Screens/ProfilePage/View/SubViews/MainView.swift]: Done

File: Instagram-like-trainee/Screens/ProfilePage/View/ProfileView.swift [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+    init(viewModel: ProfileViewModel) {
+        self.viewModel = viewModel
+    }
+
+    var body: some View {
+        VStack {
+            if let profileData = viewModel.data {
+                MainView(profileData: profileData, closeProfile: viewModel.closeProfile)
+            } else {
+                ProgressView("Loading...")
```
[the remark this answers is not in the document]
Comment (mlchct) [reply Instagram-like-trainee/Screens/ProfilePage/View/ProfileView.swift]: Done

File: Instagram-like-trainee/Screens/ProfilePage/View/SubViews/MainView.swift [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+                HeaderView(profileImage: profileData.profileImage)
+                DescriptionView(profileName: profileData.profileName, description: profileData.description)
+                HStack(spacing: 5) {
+                    createButton(name: R.string.localizable.follow(), color: .white)
+                    createButton(name: R.string.localizable.message(), color: .black)
+                    createButton(name: R.string.localizable.email(), color: .black)
+                }
+                .padding(3)
+                PostFeedView(posts: profileData.posts)
+
```
[the remark this answers is not in the document]
Comment (mlchct) [reply Instagram-like-trainee/Screens/ProfilePage/View/SubViews/MainView.swift:32]: Done

**Pull request merged**